core/src/main/scala/cats/data/HashMap.scala

```diff
@@ -0,0 +1,1196 @@
+   * according to [[cats.Hash]] instances.
+   *
+   * @param hashValue the [[cats.kernel.Hash]] instance to use for hashing values of type `VV`.
+   * @return a hash code value for this map.
+   */
+  final def hash[VV >: V](implicit hashValue: Hash[VV]): Int =
+    StaticMethods.unorderedHash(this.iterator: Iterator[(K, VV)])
+
+  final override def hashCode(): Int = {
+    implicit val valueHash = Hash.fromUniversalHashCode[V]
+    StaticMethods.unorderedHash(this.iterator)
+  }
+
+  /**
+   * Typesafe stringification operator.
+   *
+   * This method is similar to [[java.lang.Object#toString]] except that it stringifies values according
+   * to [[cats.Show]] instances.
+   *
+   * @param showKey the [[cats.Show]] instance to use for showing keys of type `K`.
+   * @param showValue the [[cats.Show]] instance to use for showing values of type `V`.
+   * @return a [[java.lang.String]] representation of this map.
+   */
+  final def show[VV >: V](implicit showKey: Show[K], showValue: Show[VV]): String =
+    iterator.map { case (k, v) => s"${showKey.show(k)} -> ${showValue.show(v)}" }.mkString("HashMap(", ", ", ")")
+
+  final override def toString() =
+    iterator.map { case (k, v) => s"$k -> $v" }.mkString("HashMap(", ", ", ")")
+}
+
+object HashMap extends HashMapInstances {
+  final private[data] def improve(hash: Int): Int =
+    scala.util.hashing.byteswap32(hash)
+
+  /**
+    * Creates a new empty [[cats.data.HashMap]] which uses `hashKey` for hashing.
+    *
+    * @param hashKey the [[cats.kernel.Hash]] instance used for hashing keys.
+    * @return a new empty [[cats.data.HashMap]].
+    */
+  final def empty[K, V](implicit hashKey: Hash[K]): HashMap[K, V] =
+    new HashMap[K, V](Node.empty[K, V])
+
+  /**
+  * Creates a new [[cats.data.HashMap]] which contains all elements of `kvs`.
+  *
+  * @param kvs the key-value pairs to add to the [[cats.data.HashMap]].
+  * @param hashKey the [[cats.kernel.Hash]] instance used for hashing keys.
+  * @return a new [[cats.data.HashMap]] which contains all elements of `kvs`.
+  */
+  final def apply[K, V](kvs: (K, V)*)(implicit hashKey: Hash[K]) =
+    fromSeq(kvs)
+
+  /**
+  * Creates a new [[cats.data.HashMap]] which contains all elements of `seq`.
+  *
+  * @param seq the sequence of elements to add to the [[cats.data.HashMap]].
+  * @param hashKey the [[cats.kernel.Hash]] instance used for hashing values.
+  * @return a new [[cats.data.HashMap]] which contains all elements of `seq`.
+  */
+  final def fromSeq[K, V](seq: Seq[(K, V)])(implicit hashKey: Hash[K]): HashMap[K, V] = {
+    val rootNode = seq.foldLeft(Node.empty[K, V]) { case (node, (k, v)) =>
+      node.add(k, improve(hashKey.hash(k)), v, 0)
+    }
+    new HashMap(rootNode)
+  }
+
+  sealed abstract private[data] class Node[K, +V] {
+
+    /**
+      * @return The number of value and node elements in the contents array of this trie node.
+      */
+    def allElements: Int
```

Comment:
nit: I would call this `sizeAllElements` similarly below. `allElements` sounds like you are getting a collection not an integer back.

Reply:
Renamed this to `allElementsCount`, and I renamed the other size fields to `keyValueCount` and `nodeCount` as well